### PR TITLE
+ #21 Add LabelSanitizer in order to survive : in label names (e.g. from Vapor)

### LINKED ISSUE #21: Consider being defensive wrt. labels which contain `:`
This is a bug which breaks using swift-statsd with Vapor.

It's debatable where this should be fixed (if label sanitization could be part of swift metrics APIs or not), but following prior discussions and art (how we fixed it in  swift-prometheus via "LabelSanitizers" https://github.com/MrLotU/SwiftPrometheus/commit/938b2a29be42e756959175315d87c978d9f8e64e ), let's perhaps fix it in the backend here.

--- 

Details:

Vapor has built in metrics and emits them here https://github.com/vapor/vapor/blob/633b2a81f547383aa6becb6ba1ad011dc7174779/Sources/Vapor/Responder/DefaultResponder.swift#L83

It does so for each route (very good):

```
            self.updateMetrics(
                for: request,
                path: path,
                startTime: startTime,
                statusCode: status.code
            )
```

and it identifies them by the route description (the problem):

```
            path = cachedRoute.route.description
```

since `route.description` would be something like: `/users/:uuid`

thus the reported metric would be:

```swift
/users/:uuid:111|c
// MetricName:value|type|@sample_rate|#tag1:
```

so the `:` of the uuid breaks the format.

---

Long story short, we need label sanitizers the same way as swift-prom does.

I'll PR one here and we can at some point consider if it's a general concept that should exist for all swift-metrics backends.

### PR DESCRIPTION
**Pull request by ktoso**
More details in the ticket.

This addresses label sanitization in the same manner as we decided back then to do it for prometheus ( https://github.com/MrLotU/SwiftPrometheus/commit/938b2a29be42e756959175315d87c978d9f8e64e ).

It allows Vapor apps to be used by default with swift-metrics + swift-statsd-client.

Resolves #21

### CODE REVIEW

File: Sources/StatsdClient/StatsdClient.swift
Comment (Lukasa) [inline Sources/StatsdClient/StatsdClient.swift]: Do we ever forsee needing a second one of these?

Comment (ktoso) [reply Sources/StatsdClient/StatsdClient.swift]: They come in handy when "library x" uses keys which read badly when escaped but you have a good idea how you'd like to represent them. (Originally I wanted to "drop" any `:`, but then `label:label` would read very badly, so defaulted to replacing with `_`)

Specifically in this example there's a project which uses Vapor and Vapor today uses labels such as: 

`/hello/:who` so the `:` breaks things, and while our "safe default substitution" to `/hello/_who` reads "ok" users may prefer to have those show up as `/hello/who` in their metrics backends or something similar. Thus leaving it open.

This also follows prior art as the Prometheus metrics backend is doing; We also debated there if it's necessary, but I guess consistency is a nice thing to have for those.

Comment (ktoso) [reply Sources/StatsdClient/StatsdClient.swift]: Do you think we should be worried about the performance overhead? In my use cases metrics objects are created and reused so it's less worrying if sanitizing the label has to call through a protocol... but perhaps there's users which "always" would `Counter("silly:label")`...

Comment (Lukasa) [reply Sources/StatsdClient/StatsdClient.swift]: I'm not so much worried about the performance overhead as I am worried about overgeneralising. Defining protocols where they aren't needed is awkward.

That said, given that there's prior art, I'm not going to stand in the way of this. That said, I do think we have a problem with the protocol names. This is the second module that defines `LabelSanitizer` as a public name, meaning that anyone who imports both will get ambiguous type errors if they attempt to conform. That's annoying. If we think this is a pattern all the metrics libraries should follow, we should arrange to ensure that part of the pattern is that these protocols all have different names!

Probably this should be called `StatsdLabelSanitizer`.

Comment (ktoso) [reply Sources/StatsdClient/StatsdClient.swift]: I see, yeah that's a good point about the names. While I'm not really expecting systems to use multiple metrics backends, there always could be some legitimate reason for doing so.

I like the naming proposal, so it'll be `StatsdLabelSanitizer` the protocol and `DefaultStatsdLabelSanitizer` the impl.

Comment (ktoso) [reply Sources/StatsdClient/StatsdClient.swift]: Alternatively we could call them `StatsdMetricNameSanitizer`, technically that's what the field this sanitizes in statsd wording is. 

The only difference I can see is a slight semantic one:
- label is label from swift-metrics
- we _also_ have to sanitize the entire metric name
  - the entire metric name equals label + dimensions

So it would be a bit more correct to call it `StatsdMetricNameSanitizer`... Maybe Tom will have an opinion on this 🤔

Comment (tomerd) [reply Sources/StatsdClient/StatsdClient.swift]: if must use a protocol I would name it `StatsdMetricNameSanitizer` and add a `NameSanitizer`  (or similar) type alias on the `StatsdClient` to make it easier on the eye.

Since this is a single function I would also consider dropping the protocol and use a function instead, e.g.:

```swift
public final class StatsdClient: MetricsFactory  {
   typealias NameSanitizer = (String) -> String
   ...
}
```

Comment (ktoso) [reply Sources/StatsdClient/StatsdClient.swift]: Sounds good, simplified here: https://github.com/apple/swift-statsd-client/pull/22/commits/2910f13b63d089e6ba88df13e500d02bf82edbeb

It's just a typealias now.

I just added the (defaulted) parameter to the existing init since we're in "alpha" versions.

**Comment by ktoso** [comment]
@tomerd (can't merge without an LGTM :-)).

I can also tag the release but need super-powers to the repo then 🙏

**Review by tomerd — APPROVED** [review]

**Pull request merged**
